runtime/vm/JVMImage.cpp

```diff
@@ -0,0 +1,103 @@
+/*******************************************************************************
+ * Copyright (c) 2001, 2019 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#include "JVMImage.hpp"
+
+#define OMRPORT_FROM_IMAGE JVMImage::getJVMImage()->getPortLibrary();
+
+JVMImage::JVMImage(J9JavaVM *javaVM) :
+	_memoryStart(0),
+	_size(0),
+	_isImageAllocated(false),
+	_portLibrary(NULL)
+{
+	OMRPORT_ACCESS_FROM_J9PORT(javaVM->portLibrary);
+	memcpy(_portLibrary, privateOmrPortLibrary, sizeof(OMRPortLibrary));
+	_portLibrary->mem_allocate_memory = image_mem_allocate_memory;
```

Comment:
mem_allocate_memory is defined as `void  *(*mem_allocate_memory)(struct OMRPortLibrary *portLibrary, uintptr_t byteAmount, const char *callSite, uint32_t category) ;` so youll need to assign it to a function with the same contract

You could modify the existing one to just be `image_mem_allocate_memory(uintptr_t byteAmount)` and then make another `image_port_mem_allocate_memory(struct OMRPortLibrary *portLibrary, uintptr_t byteAmount, const char *callSite, uint32_t category)` which calls the previous one.

Most of the time you will just need the `image_mem_allocate_memory` version.

Reply:
wait I thought `image_mem_allocate_memory` has the same definition as `mem_allocate_memory`. I looked at the class loader initialization  which involves pool and hashtable and they both rely on `mem_allocate_memory`. I could see having a separate `image_mem_allocate_memory(uintptr_t byteAmount)` being beneficial if there is a place I allocate where I can not pass in OMRPortLib

Reply:
If you look at j9port_generated.h j9mem_allocate_memory used in the VM is defined as:
`#define j9mem_allocate_memory(param1, category) OMRPORT_FROM_J9PORT(privatePortLibrary)->mem_allocate_memory(OMRPORT_FROM_J9PORT(privatePortLibrary),param1, J9_GET_CALLSITE(), category)`

So the version of image_mem_allocate that is passed to the portlib needs the full param list.

> image_mem_allocate_memory(uintptr_t byteAmount) being beneficial if there is a place I allocate where I can not pass in OMRPortLib

Yes, this will be most of the usages in the VM.

Reply:
My bad what I meant was my function has the same definition as `portLib->mem_allocate_memory`. See line 89. As for the wrapper around this function that only takes in a int size I will put that in. My plan for the class loaders at least was to change keyHashTable.c : hashClassTableNew's portlib passed in with the macro I defined and change jvminit.c:2057 instead of passing pool_for_port pass in the mem_alloc functions. But I haven't looked at the rest yet (classpath, class)

Reply:
Im going to resolve this comment right now so the other devs could continue their branch, but Ill come by your desk tomorrow morning to resolve this comment and put in a fix pr.